python_modules/dagster/dagster/core/events/execution.py

```diff
@@ -0,0 +1,158 @@
+from collections import namedtuple
+from enum import Enum
+
+from dagster import check
+from dagster.core.execution_plan.objects import (
+    StepKind,
+    StepOutputData,
+    StepFailureData,
+    StepSuccessData,
+)
+from dagster.core.execution_context import SystemStepExecutionContext
+from .logging import EventType
+
+
+class ExecutionStepEventType(Enum):
+    STEP_OUTPUT = 'STEP_OUTPUT'
+    STEP_FAILURE = 'STEP_FAILURE'
+    STEP_START = 'STEP_START'
+    STEP_SUCCESS = 'STEP_SUCCESS'
+
+
+class ExecutionStepEvent(
+    namedtuple(
+        '_ExecutionStepEvent',
+        'event_type_value step_key pipeline_name solid_name solid_definition_name step_kind_value '
+        'step_output_data step_failure_data step_success_data tags',
+    )
+):
+    @staticmethod
+    def from_step(event_type, step_context, step_output_data, step_failure_data, step_success_data):
+        event = ExecutionStepEvent(
+            check.inst_param(event_type, 'event_type', ExecutionStepEventType).value,
+            step_context.step.key,
+            step_context.pipeline_def.name,
+            step_context.step.solid.name,
+            step_context.step.solid.definition.name,
+            step_context.step.kind.value,
+            step_output_data,
+            step_failure_data,
+            step_success_data,
+            step_context.tags,
+        )
+        step_context.log.info(
+            ('{event_type} for step {step_key}').format(
+                event_type=event_type, step_key=step_context.step.key
+            ),
+            event_type=EventType.DAGSTER_EVENT.value,
+            dagster_event=event,
+            pipeline_name=step_context.pipeline_def.name,
+        )
+        return event
+
+    def __new__(
+        cls,
+        event_type_value,
+        step_key,
+        pipeline_name,
+        solid_name,
+        solid_definition_name,
+        step_kind_value,
+        step_output_data,
+        step_failure_data,
+        step_success_data,
+        tags,
+    ):
+        return super(ExecutionStepEvent, cls).__new__(
+            cls,
+            check.str_param(event_type_value, 'event_type_value'),
+            check.str_param(step_key, 'step_key'),
+            check.str_param(pipeline_name, 'pipeline_name'),
+            check.str_param(solid_name, 'solid_name'),
+            check.str_param(solid_definition_name, 'solid_definition_name'),
+            check.str_param(step_kind_value, 'step_kind_value'),
+            check.opt_inst_param(step_output_data, 'step_output_data', StepOutputData),
```

Comment:
no need to do this now, but i think we will want to change step_output_data, step_failure_data, and step_success_data to be one `event_specific_data` property that is a union.